File: leetCode/string/easy/quiz125/java/SolutionV1.java
package string.easy.quiz125.java;

public class SolutionV1 {
    public static void main(String[] args) {
        boolean answer = new SolutionV1().isPalindrome("0P");
        System.out.println(answer);
    }

    public boolean isPalindrome(String s) {
        String cleanedInput = doCleaning(s);
        int n = cleanedInput.length();

        for (int i = 0; i < n / 2; i++) {
            if (cleanedInput.charAt(i) != cleanedInput.charAt(n - 1 - i)) {
                return false;
            }
        }

        return true;
    }

    private String doCleaning(String s) {
        StringBuilder sb = new StringBuilder();
        for (char c : s.toCharArray()) {
            if ((c - 'a' < 26 && c - 'a' >= 0) || (c - 'A' < 26 && c - 'A' >= 0) || (c > 47 && c < 58)) {
                sb.append(c);
            }
        }
        return sb.toString().toLowerCase();
    }
}

File: leetCode/string/easy/quiz125/kotlin/SolutionV1.kt
package string.easy.quiz125.kotlin

fun main() {
    SolutionV1().isPalindrome("A man, a plan, a canal: Panama").apply {
        println(this)
    }
}

class SolutionV1 {
    fun isPalindrome(s: String): Boolean {
        val cleanedInput = s.doCleaning();
        val n = cleanedInput.length

        for (left in 0 until n / 2) {
            val right = n - 1 - left
            if (cleanedInput[left] != cleanedInput[right]) return false
        }
        return true
    }

    private fun String.doCleaning(): String {
        return with(StringBuilder()) {
            for (c in this@doCleaning) {
                if (Character.isLetterOrDigit(c)) append(c)
            }
            toString().lowercase()
        }
    }
}
